python/lsst/daf/butler/core/dimensions/elements.py

```diff
@@ -0,0 +1,194 @@
+    """Base class for elements in the dimension schema.
+
+    `Dimension` objects are not directly constructable; they can only be
+    obtained from a `DimensionGraph`.
+    """
+
+    def __init__(self, universe, name, hasRegion, link, required, optional, doc):
+        self._universe = universe
+        self._name = name
+        self._hasRegion = hasRegion
+        self._link = frozenset(link)
+        self._doc = doc
+        self._requiredDependencies = DimensionUnitSet(universe, required, optional=False)
+        self._allDependencies = DimensionUnitSet(universe, set(required) | set(optional), optional=True)
+        # Compute _primaryKeys dict, used to back primaryKeys property.
+        primaryKey = set(self.link)
+        for dimension in self.dependencies(optional=False):
+            primaryKey |= dimension.link
+        self._primaryKey = frozenset(primaryKey)
+
+    def __eq__(self, other):
+        try:
+            return self.universe is other.universe and self.name == other.name
+        except AttributeError:
+            return NotImplemented
+
+    def __hash__(self):
+        return hash(tuple(self.name))
+
+    @property
+    def universe(self):
+        """The graph of all dimensions compatible with self (`DimensionGraph`).
+        """
+        return self._universe
+
+    @property
+    def name(self):
+        """Name of this dimension (`str`, read-only).
+
+        Also assumed to be the name of the associated table (if present).
+        """
+        return self._name
+
+    @property
+    def hasRegion(self):
+        """Whether this dimension is associated with a region on the sky
+        (`bool`).
+        """
+        return self._hasRegion
+
+    @property
+    def doc(self):
+        """Documentation for this dimension (`str`).
+        """
+        return self._doc
+
+    @property
+    def primaryKey(self):
+        """The names of fields that uniquely identify this dimension in a
+        data ID dict(`frozenset` of `str`).
+        """
+        return self._primaryKey
+
+    @property
+    def link(self):
+        """Primary key fields that are used only by this dimension, not any
+        dependencies (`frozenset` of `str`).
+        """
+        return self._link
+
+    def dependencies(self, optional=True):
+        """Return the set of dimension units this dimension depends on.
+
+        Parameters
+        ----------
+        optional : `bool`
+            If `True` (default), include optional as well as required
+            dependencies. The primary keys of required dependencies are always
+            included in the primary key of the dimension that depends on them,
+            while the primary keys of optional dependencies are not.
+
+        Returns
+        -------
+        dependencies : `DimensionSet` of `DimensionUnit`
+        """
+        return self._allDependencies if optional else self._requiredDependencies
+
+
+class DimensionUnit(Dimension):
```

Comment:
Agreed.

Reply:
I gave this a try, and it doesn't quite work; the names for lots of other attributes and methods make more sense if the name for the base concept is really a first-class name.

Would `elements` instead of `units` be any better?  Any other ideas?